File: android/src/main/java/com/scandit/capacitor/datacapture/core/utils/CapacitorResult.kt
package com.scandit.capacitor.datacapture.core.utils

import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.scandit.datacapture.frameworks.core.result.FrameworksResult
import org.json.JSONObject

class CapacitorResult(private val call: PluginCall) : FrameworksResult {
    override fun success(result: Any?) {
        if (result is Map<*, *>) {
            call.resolve(JSObject.fromJSONObject(JSONObject(result)))
        } else if (result != null) {
            call.resolve(JSObject().putSafe("data", result))
        } else {
            call.resolve()
        }
    }

    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
        call.reject(errorMessage, errorMessage)
    }
}

File: ios/Plugin/CapacitorResult.swift
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2023- Scandit AG. All rights reserved.
 */

import Capacitor
import ScanditFrameworksCore

public struct CapacitorResult: FrameworksResult {
    private let pluginCall: CAPPluginCall

    public init(_ call: CAPPluginCall) {
        pluginCall = call
    }

    public func success(result: Any?) {
        if let result = result as? [String: Any] {
            pluginCall.resolve(result)
        } else if let result = result {
            pluginCall.resolve(["data": result])
        } else {
            pluginCall.resolve()
        }
    }
    
    public func reject(code: String, message: String?, details: Any?) {
        pluginCall.reject(message!, code)
    }
    
    public func reject(error: Error) {
        let message = String(describing: error)
        let nsError = error as NSError
        pluginCall.reject(message, "\(nsError.code)", error)
    }
}
